### ISSUE TITLE
test/integration/mallocx failures on Windows

### ISSUE THREAD

**Issue by glandium** [jemalloc/jemalloc#393]
test_basic fails with:
```
test_basic:test/integration/mallocx.c:125: Failed assertion: (p) != (NULL) --> 0x0 == 0x0: Unexpected mallocx() error
```
test_alignment_and_size fails with:
```
test/integration/mallocx.c:177: Failed assertion: (ps[i]) != (NULL) --> 0x0 == 0x0: mallocx() error for alignment=4194304, size=4194303 (0x3fffff)
```

**Comment by jasone** [comment]
I don't currently see any test failures on Windows (using MinGW).  How can I reproduce this?

**Comment by glandium** [comment]
I built with MSVC (I don't remember if it was 2013 or 2015).

**Comment by glandium** [comment]
I've set up Windows builds on AppVeyor (MINGW32, MINGW64), and they all fail in different ways:

MINGW32:
```
test_alignment_and_size:test/integration/mallocx.c:175: Failed assertion: (ps[i]) != (NULL) --> 0x0 == 0x0: mallocx() error for alignment=33554432, size=83886076 (0x4fffffc)
test/test.sh: line 22:  2828 Segmentation fault      ${t}.exe /c/projects/jemalloc/ /c/projects/jemalloc/
Test harness error
```

MINGW64:
```
[00:08:56] test_quarantine:test/unit/quarantine.c:51: Failed assertion: (p) != (quarantined[j]) --> 0x3a1000 == 0x3a1000: Quarantined region recycled too early; i=1, j=0
```
(for all(?) values of i and j)
```
[00:09:01] test_zero:test/unit/zero.c:33: Failed assertion: (s[i]) == (0x0) --> 97 != 0: Newly allocated byte 0/65536 isn't zero-filled
```
(for all(?) values from 0/65536 to 57343/65536)

I'm also trying to get MSVC builds running.

https://ci.appveyor.com/project/glandium/jemalloc/build/4

**Comment by glandium** [comment]
Now with, in the order they appear:
- MSVC 2015 64-bits
- MSVC 2015 32-bits
- MINGW64
- MINGW32

Only one of them passes all checks (MSVC 2015 64-bits).

https://ci.appveyor.com/project/glandium/jemalloc/build/28

**Comment by glandium** [comment]
This may or may not be related, but 32-bits Linux and OSX fail in test/integration/mallocx too:
https://travis-ci.org/glandium/jemalloc/builds/136479829

**Comment by glandium** [comment]
This is failing on the rel-4.3.0 branch too.

**Comment by jasone** [comment]
871a9498e13572f99451ed88db36cda6c9fecf8f fixes some overflow bugs on 32-bit CentOS 6.  Those bugs do not appear to be the ones reported here, but may be related.

**Comment by jasone** [comment]
I think the reason for failure is that the ```malloc_conf``` weak symbol isn't being overridden.  I tried to fix this in 7b24c6e5570062495243f1e55131b395adb31e33 / 5569b4a42c85f951f783b315bd49668a6fa764e3 (and verified locally that it fixes a failure in test/integration/chunk), but according to AppVeyor it needs to be conditionalized on gcc, and even worse, it causes new segfaults for gcc-compiled tests.

**Comment by jasone** [comment]
There appear to be three related causes:
- Memory pressure due to allocating over-sized rtree leaf nodes (fixed; dev: 68e14c988460907c4b135feb3eb5fccd28953feb, rel-4.3.0: dc553d52d82380ab0c99bb0fa97ae5c1f11ab2be)
- Inability to merge virtual memory mappings on Windows, in combination with map fragmentation caused by the rtree leaves.
- Increased retained memory as a result of decay-based purging (dev-specific).

Assuming this is the full set of causes, I think this will be straightforward to finish resolving via some tweaks to the tests that reduce virtual memory usage.

**Comment by jasone** [comment]
The ```mallocx()``` failures appear to be fully resolved now.
